src/distribution/reference.rs

```diff
@@ -0,0 +1,449 @@
+    pub fn tag(&self) -> Option<&str> {
+        self.tag.as_deref()
+    }
+
+    /// Returns the object's digest, if present.
+    pub fn digest(&self) -> Option<&str> {
+        self.digest.as_deref()
+    }
+
+    /// Returns the original registry when pulled via a mirror.
+    ///
+    /// Since this is not part of the stable OCI Distribution Spec yet, this feature is exempt from
+    /// semver backwards compatibility guarantees and might change in the future.
+    #[doc(hidden)]
+    pub fn namespace(&self) -> Option<&str> {
+        if self.mirror_registry.is_some() {
+            Some(self.registry())
+        } else {
+            None
+        }
+    }
+
+    /// Returns the full repository name and path.
+    fn full_name(&self) -> String {
+        if self.registry() == "" {
+            self.repository().to_string()
+        } else {
+            format!("{}/{}", self.registry(), self.repository())
+        }
+    }
+
+    /// Returns the whole reference.
+    pub fn whole(&self) -> String {
+        let mut s = self.full_name();
+        if let Some(t) = self.tag() {
+            if !s.is_empty() {
+                s.push(':');
+            }
+            s.push_str(t);
+        }
+        if let Some(d) = self.digest() {
+            if !s.is_empty() {
+                s.push('@');
+            }
+            s.push_str(d);
+        }
+        s
+    }
+}
+
+impl fmt::Display for Reference {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.whole())
+    }
+}
+
+impl FromStr for Reference {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Reference::try_from(s)
+    }
+}
+
+impl TryFrom<String> for Reference {
+    type Error = ParseError;
+
+    fn try_from(s: String) -> Result<Self, Self::Error> {
+        if s.is_empty() {
+            return Err(ParseError::NameEmpty);
+        }
+        let captures = match reference_regexp().captures(&s) {
+            Some(caps) => caps,
+            None => {
+                return Err(ParseError::ReferenceInvalidFormat);
+            }
+        };
+        let name = &captures[1];
+        let mut tag = captures.get(2).map(|m| m.as_str().to_owned());
+        let digest = captures.get(3).map(|m| m.as_str().to_owned());
+        if tag.is_none() && digest.is_none() {
+            tag = Some(DEFAULT_TAG.into());
+        }
+        let (registry, repository) = split_domain(name);
+        let reference = Reference {
+            registry,
+            mirror_registry: None,
+            repository,
+            tag,
+            digest,
+        };
+        if reference.repository().len() > NAME_TOTAL_LENGTH_MAX {
+            return Err(ParseError::NameTooLong);
+        }
+        // Digests much always be hex-encoded, ensuring that their hex portion will always be
+        // size*2
+        if let Some(digest) = reference.digest() {
+            match digest.split_once(':') {
```

Comment:
This could also clearly use https://github.com/containers/oci-spec-rs/commit/5601907d5c4ead6ae42ea191f485726fd6179f98 etc. but probably best to keep code changes minimal until we've deduplicated.

Reply:
Yep I agree here. We just had this as a string for simplicity, but it would probably be better to have it be a concrete digest type. Definitely going to be easier to wait until this is merge though